lib/os/ring_buffer.c

```diff
@@ -21,6 +29,59 @@ struct ring_element {
 	uint32_t  value  :8;  /**< Room for small integral values */
 };
 
+static uint32_t mod(struct ring_buf *buf, uint32_t val)
+{
+	return likely(buf->mask) ? val & buf->mask : val % buf->size;
+}
+
+/* Check if indexes did not progresses to far (too close to 32 bit wrapping).
```

Comment:
Check if indexes did not progress too far (too close to 32-bit wrapping). If so, then reduce all indexes by an arbitrary value.